src/stan/math/prim/scal/fun/value_of_rec.hpp

```diff
@@ -12,33 +12,14 @@ namespace stan {
      * <p>See the <code>stan::math::primitive_value</code> function to 
      * extract values without casting to <code>double</code>.
      *
-     * <p>This function is meant to cover the primitive types. For
-     * types requiring pass-by-reference, this template function
-     * should be specialized.
-     *
-     * @tparam T Type of scalar.
      * @param x Scalar to convert to double.
      * @return Value of scalar cast to a double.
      */
-    template <typename T>
-    inline double value_of_rec(const T x) {
-      return static_cast<double>(x);
+    inline double value_of_rec(double x) {
```

Comment:
Ah yes, this is a problem that I ran into in a few different places. The problem is that template<typename T> matches when T=Eigen::MatrixBase<T1> even when there is a more specific template<typename T> for Eigen::MatrixBase. I couldn't figure out how to get the template matching to work without removing the most generic template<typename T>. Maybe I am missing something?